### PR TITLE
Using nrnivmodl-core to build new neurodamus packages

### PR DESCRIPTION
**Pull request by ferdonline**
This PR implements the building of neurodamus (new organization) using nrnivmodl-core and thus ~~dropping coreneuron-xxx~~ (we keep for backward compat).
Building on this PR, we can activate tests using coreneuron with the same exact mod files as those coming from the new neurodamus models. 
Besides, we now use the "soft-linked" mods of the model. Besides being simpler for the user to build stand-alone models, beuilding neurodamus-<model> is also easier since only neurodamus-core has to be brought in.

E.g. When the user requests the `neurodamus-neocortex` spack package he can specify `+coreneuron+plasticity` which will take the new neocortex mods with plasticity changes and build the mod files against neuron and coreneuron, in a final setup similar to the previous plasticity branch. However only a coreneuron instance is required in the system, akin to neuron.

### CODE REVIEW

**Comment by ferdonline** [comment]
Are there changes required in this PR?

File: var/spack/repos/builtin/packages/neurodamus-core/package.py [hunk 1]
```diff
@@ -14,7 +14,7 @@ class NeurodamusCore(Package):
     version('develop', git=git, branch='master')
     version('2.0.0', tag='2.0.0')
 
-    variant('python', default=True, description="Enable Python Neurodamus")
+    variant('python', default=False, description="Enable Python Neurodamus")
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-core/package.py:16]: we can also install this by default?

File: var/spack/repos/builtin/packages/neurodamus-hippocampus/package.py [hunk 2]
```diff
@@ -12,5 +12,4 @@ class NeurodamusHippocampus(NeurodamusModel):
     git      = "ssh://bbpcode.epfl.ch/sim/models/hippocampus"
 
     version('develop', git=git, branch='master', submodules=True)
-    version('1.0', tag='1.0', submodules=True)
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-hippocampus/package.py:15]: If tests are passing, we should tag this as version...eg 0.1?

File: var/spack/repos/builtin/packages/neurodamus-model/package.py [hunk 3]
```diff
@@ -11,20 +11,21 @@ class NeurodamusModel(Package):
...
-
     variant('coreneuron',  default=False, description="Enable CoreNEURON Support")
     variant('profile',     default=False, description="Enable profiling using Tau")
     variant('synapsetool', default=True,  description="Enable Synapsetool reader")
     variant('sonata',      default=False, description="Enable Synapsetool with Sonata")
     variant('plasticity',  default=False, description="Use optimized ProbAMPANMDA_EMS and ProbGABAAB_EMS")
+    variant('python',      default=False, description="Install neurodamus-python alongside")
+
+    depends_on('neurodamus-core')
+    depends_on('neurodamus-core+python', when='+python')
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-model/package.py:21]: Fernando, just to keep things simple I wonder if we have to add python as variant. We just install by default but not announce this to user.

File: var/spack/repos/builtin/packages/neurodamus-model/package.py [hunk 4]
```diff
@@ -50,79 +51,105 @@ class NeurodamusModel(Package):
...
+
+        # If we shall build mods for coreneuron, only bring from core those specified
+        if spec.satisfies("+coreneuron"):
+            mkdirp('core_mechs')
+            with open(core_prefix.mod.join("coreneuron_modlist.txt")) as core_mods:
+                for aux_mod in core_mods:
+                    shutil.copy(core_prefix.mod.join(aux_mod.strip()), 'core_mechs')
 
         if spec.satisfies('+plasticity'):
-            self.copy_all('common/mod/optimized', 'common/mod')
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-model/package.py:66]: These are different versions of synapses right? We should just change the name from optimised to modified synapses?

File: var/spack/repos/builtin/packages/neurodamus-model/package.py [hunk 5]
```diff
@@ -50,79 +51,105 @@ class NeurodamusModel(Package):
...
-        copy_tree(spec['neurodamus-core'].prefix.mod, '_merged_mod')
+        copy_tree(core_prefix.hoc, '_merged_hoc')
+        copy_tree(core_prefix.mod, '_merged_mod')
+
+        # If we shall build mods for coreneuron, only bring from core those specified
+        if spec.satisfies("+coreneuron"):
+            mkdirp('core_mechs')
+            with open(core_prefix.mod.join("coreneuron_modlist.txt")) as core_mods:
+                for aux_mod in core_mods:
+                    shutil.copy(core_prefix.mod.join(aux_mod.strip()), 'core_mechs')
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-model/package.py]: May be we can look today at these mod files and avoid this step by making mod files coreneuron compatible?

File: var/spack/repos/builtin/packages/neurodamus-model/package.py (hunk 4 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py:66]: Optimized are the plasticity mods

**Comment by ferdonline** [comment]
@pramodk tests are now passing. You think we should create tags?
Tests are running again now activating soma report in plasticity. Something is fishy there.

**Comment by ferdonline** [comment]
Soma reports are different when using coreNeuron. I thought it could be related to the order of the cells instantiation but with the latest patch in neurodamus-core it will use RR also in CoreNeuron.
It might still be the fact that we use a sub-target. I will comment the soma report for this test for the moment.
To my mind we should solve these small issues before tagging

File: var/spack/repos/builtin/packages/neurodamus-model/package.py (hunk 3 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py:21]: You mean for the moment we dont even show the option?

File: var/spack/repos/builtin/packages/neurodamus-core/package.py (hunk 1 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-core/package.py:16]: As discussed, I still want to test neurodamus-py better.
Other issues include: 
 - Have python dependencies all available in spack streams
 - implement (at least some) of the missing features compared to hoc

File: var/spack/repos/builtin/packages/neurodamus-hippocampus/package.py (hunk 2 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-hippocampus/package.py:15]: I would tag only when all tests are passing and we solved the unknown things. Besides now that we use LC_ALL='C' we will need changes either to recipe or results

File: var/spack/repos/builtin/packages/neurodamus-model/package.py (hunk 5 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py]: Is it worth to change? I think in the medium term coreneuron should be able simply to skip those files which are not important

File: var/spack/repos/builtin/packages/neurodamus-model/package.py [hunk 6]
```diff
@@ -150,8 +183,45 @@ def setup_environment(self, spack_env, run_env):
...
+    if os.path.islink(dst):
+        os.remove(dst)
+    os.symlink(src, dst)
+
+
+def filter_out(src, dst):
+    """Remove src from dst, copy compat"""
+    fname = join_path(dst, os.path.basename(src))
+    if os.path.exists(fname):
+        os.remove(join_path(dst, os.path.basename(src)))
```
Comment (ferdonline) [inline var/spack/repos/builtin/packages/neurodamus-model/package.py]: something went wrong where
```suggestion
        os.remove(fname)
```

File: var/spack/repos/builtin/packages/neurodamus-core/package.py (hunk 1 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/neurodamus-core/package.py:16]: ok understood. fine.

File: var/spack/repos/builtin/packages/neurodamus-hippocampus/package.py (hunk 2 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/neurodamus-hippocampus/package.py:15]: Ok!

File: var/spack/repos/builtin/packages/neurodamus-model/package.py (hunk 3 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py:21]: its ok...I was thinking if python can be always installed...no variant exposed...but its ok for now.

File: var/spack/repos/builtin/packages/neurodamus-model/package.py (hunk 5 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py]: ok..I dont want to delay the merge...go ahead with current setup.

**Comment by pramodk** [comment]
Ok...lets check about soma reports tomorrow. We should be able to get some help from @jplanasc to identify the issue.

File: var/spack/repos/builtin/packages/neurodamus-model/package.py [hunk 7]
```diff
@@ -45,84 +44,107 @@ class NeurodamusModel(Package):
...
 
         # Copy from the several sources
         for hoc_src in self._hoc_srcs:
-            self.copy_all(hoc_src, '_merged_hoc')
+            copy_all(hoc_src, '_merged_hoc')
         for mod_src in self._mod_srcs:
-            self.copy_all(mod_src, '_merged_mod')
+            copy_all(mod_src, '_merged_mod')
+            if spec.satisfies("+coreneuron"):
+                copy_all(mod_src, 'core_mechs')
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-model/package.py]: Isn't this now unnecessary with symlinks?

File: var/spack/repos/builtin/packages/neurodamus-model/package.py [hunk 8]
```diff
@@ -131,13 +153,24 @@ def install(self, spec, prefix):
...
-        if os.path.isdir(py_src):
-            os.makedirs(prefix.python)
-            force_symlink('../lib', prefix.python.lib)
+        if spec.satisfies('+coreneuron'):
+            # move special-core to bin and libcorenrnmech* to lib
+            outdir = spec.architecture.target
+            shutil.move(join_path(outdir, 'special-core'), prefix.bin)
+            for libname in find_libraries("libcorenrnmech*", outdir):
+                shutil.move(libname, prefix.lib)
+            # Then modc
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-model/package.py:165]: @ferdonline  : could you post the output of `ldd special-core` after install? I wonder if paths are getting resolved if we move stuff around? (ahh...is this all static?)

File: var/spack/repos/builtin/packages/neurodamus-model/package.py [hunk 9]
```diff
@@ -150,8 +183,45 @@ def setup_environment(self, spack_env, run_env):
...
 @contextmanager
 def profiling_wrapper_on():
     os.environ["USE_PROFILER_WRAPPER"] = "1"
     yield
     del os.environ["USE_PROFILER_WRAPPER"]
+
+
+# Aux funcs
+# ---------
+def copy_all(src, dst, copyfunc=shutil.copy):
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-model/package.py:200]: with symlinks, this wont be used at all?

File: var/spack/repos/builtin/packages/neurodamus-neocortex/package.py [hunk 10]
```diff
@@ -13,12 +12,17 @@ class NeurodamusNeocortex(NeurodamusModel):
...
+    mech_name = "neocortex"
 
     @run_before('merge_hoc_mod')
-    def include_v5(self):
+    def prepare_mods(self):
         if self.spec.satisfies('+v5'):
-            self.copy_all('mod/v5', 'mod', copyfunc=shutil.move)
-
+            copy_all('mod/v5', 'mod', copyfunc=copy_all.symlink2)
+        copy_all('mod/v6', 'mod', copyfunc=copy_all.symlink2)
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-neocortex/package.py:25]: What is your thought about our discussion of building two specials in same module :  `v5/special` and `v6/special`  ?

Or, we will need to provide two modules : neurocortext/0.1/v5 and neurocortext/0.1/v6

File: var/spack/repos/builtin/packages/neurodamus-model/package.py (hunk 7 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py]: Neurodamusc-core is staying independent, no sym-links. So here we merge both.
I keep the for loop, but in the end mods only do one iteration now.

File: var/spack/repos/builtin/packages/neurodamus-model/package.py (hunk 8 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py:165]: We also use rpath to make it more solid. Outputting also readelf so you can review.

```
readelf -d /gpfs/bbp.cscs.ch/home/leite/test/blueconfigs/INSTALL_HOME/install/linux-rhel7-x86_64/intel-18.0.3/neurodamus-neocortex-develop-mc2tue/bin/special-core

Dynamic section at offset 0x47d20 contains 38 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libcorenrnmech_neocortex-develop.so]
 0x0000000000000001 (NEEDED)             Shared library: [libcoreneuron.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libreportinglib.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libhdf5.so.101]
 0x0000000000000001 (NEEDED)             Shared library: [libz.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libsyn2.so]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libmpi++abi1002.so]
 0x0000000000000001 (NEEDED)             Shared library: [/opt/hpe/hpc/mpt/mpt-2.16/lib/libmpi.so]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x000000000000000f (RPATH)              Library rpath: [/gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/intel-18.0.3-kwbptw7qxx/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64:/gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/gcc-6.4.0-i6lyqfscua/lib:/gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/gcc-6.4.0-i6lyqfscua/lib64:$ORIGIN:/gpfs/bbp.cscs.ch/home/leite/test/blueconfigs/INSTALL_HOME/install/linux-rhel7-x86_64/intel-18.0.3/coreneuron-develop-bmyahd/lib64:/gpfs/bbp.cscs.ch/home/leite/test/blueconfigs/INSTALL_HOME/install/linux-rhel7-x86_64/intel-18.0.3/neurodamus-neocortex-develop-mc2tue/lib:/gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/08-08-2018/install/linux-rhel7-x86_64/intel-18.0.1/reportinglib-develop-uasv33/lib:/gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/18-06-2018/install/linux-rhel7-x86_64/gcc-6.4.0/hdf5-1.10.2-dueuws/lib:/gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/2018-09-28/install/linux-rhel7-x86_64/gcc-6.4.0/zlib-1.2.11-w43e56/lib:/gpfs/bbp.cscs.ch/home/leite/test/blueconfigs/INSTALL_HOME/install/linux-rhel7-x86_64/intel-18.0.3/synapsetool-0.4.1-kpne43/lib64:/gpfs/bbp.cscs.ch/home/leite/test/blueconfigs/INSTALL_HOME/install/linux-rhel7-x86_64/intel-18.0.3/neurodamus-neocortex-develop-mc2tue/lib64:/gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/18-06-2018/install/linux-rhel7-x86_64/gcc-6.4.0/boost-1.67.0-6skhmz/lib:/gpfs/bbp.cscs.ch/project/proj12/jenkins/devel_builds/install/linux-rhel7-x86_64/intel-18.0.1/neuron-2018-10-3a3zr2/lib:/opt/hpe/hpc/mpt/mpt-2.16/lib]
```


```
leite@bbpv2 ~/test/blueconfigs (master %)$ ldd /gpfs/bbp.cscs.ch/home/leite/test/blueconfigs/INSTALL_HOME/install/linux-rhel7-x86_64/intel-18.0.3/neurodamus-neocortex-develop-mc2tue/bin/special-core
	linux-vdso.so.1 =>  (0x00007fffedb05000)
	libcorenrnmech_neocortex-develop.so => /gpfs/bbp.cscs.ch/home/leite/test/blueconfigs/INSTALL_HOME/install/linux-rhel7-x86_64/intel-18.0.3/neurodamus-neocortex-develop-mc2tue/lib/libcorenrnmech_neocortex-develop.so (0x00007fffed8ad000)
	libcoreneuron.so.1 => /gpfs/bbp.cscs.ch/home/leite/test/blueconfigs/INSTALL_HOME/install/linux-rhel7-x86_64/intel-18.0.3/coreneuron-develop-bmyahd/lib64/libcoreneuron.so.1 (0x00007fffed5b2000)
	libreportinglib.so.2 => /gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/08-08-2018/install/linux-rhel7-x86_64/intel-18.0.1/reportinglib-develop-uasv33/lib/libreportinglib.so.2 (0x00007fffed36c000)
	libhdf5.so.101 => /gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/18-06-2018/install/linux-rhel7-x86_64/gcc-6.4.0/hdf5-1.10.2-dueuws/lib/libhdf5.so.101 (0x00007fffecd7c000)
	libz.so.1 => /gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/2018-09-28/install/linux-rhel7-x86_64/gcc-6.4.0/zlib-1.2.11-w43e56/lib/libz.so.1 (0x00007fffecb65000)
	libsyn2.so => /gpfs/bbp.cscs.ch/home/leite/test/blueconfigs/INSTALL_HOME/install/linux-rhel7-x86_64/intel-18.0.3/synapsetool-0.4.1-kpne43/lib64/libsyn2.so (0x00007fffec87f000)
	libpthread.so.0 => /usr/lib64/libpthread.so.0 (0x00007fffec62f000)
	libmpi++abi1002.so => /opt/hpe/hpc/mpt/mpt-2.16/lib/libmpi++abi1002.so (0x00007fffec3f4000)
	/opt/hpe/hpc/mpt/mpt-2.16/lib/libmpi.so (0x00007fffec01f000)
	libstdc++.so.6 => /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/gcc-6.4.0-i6lyqfscua/lib64/libstdc++.so.6 (0x00007fffebc9d000)
	libm.so.6 => /usr/lib64/libm.so.6 (0x00007fffeb99b000)
	libgcc_s.so.1 => /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/gcc-6.4.0-i6lyqfscua/lib64/libgcc_s.so.1 (0x00007fffeb784000)
	libc.so.6 => /usr/lib64/libc.so.6 (0x00007fffeb3c0000)
	libdl.so.2 => /usr/lib64/libdl.so.2 (0x00007fffeb1bc000)
	libimf.so => /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/intel-18.0.3-kwbptw7qxx/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64/libimf.so (0x00007fffeac28000)
	libsvml.so => /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/intel-18.0.3-kwbptw7qxx/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64/libsvml.so (0x00007fffe931a000)
	libirng.so => /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/intel-18.0.3-kwbptw7qxx/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64/libirng.so (0x00007fffe8fa5000)
	libintlc.so.5 => /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/intel-18.0.3-kwbptw7qxx/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64/libintlc.so.5 (0x00007fffe8d37000)
	libsz.so.2 => /gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/18-06-2018/install/linux-rhel7-x86_64/gcc-6.4.0/libszip-2.1.1-gszkr7/lib/libsz.so.2 (0x00007fffe8b21000)
	libcpuset.so.1 => /usr/lib64/libcpuset.so.1 (0x00007fffe8914000)
	libbitmask.so.1 => /usr/lib64/libbitmask.so.1 (0x00007fffe870f000)
	libboost_filesystem.so.1.67.0 => /gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/18-06-2018/install/linux-rhel7-x86_64/gcc-6.4.0/boost-1.67.0-6skhmz/lib/libboost_filesystem.so.1.67.0 (0x00007fffe84f3000)
	libboost_system.so.1.67.0 => /gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/18-06-2018/install/linux-rhel7-x86_64/gcc-6.4.0/boost-1.67.0-6skhmz/lib/libboost_system.so.1.67.0 (0x00007fffe82ed000)
	/lib64/ld-linux-x86-64.so.2 (0x0000555555554000)
	librt.so.1 => /usr/lib64/librt.so.1 (0x00007fffe80e5000)
```

File: var/spack/repos/builtin/packages/neurodamus-model/package.py (hunk 9 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py:200]: Will be, to merge with neurodamus-core. Anyway I find this a very useful little func. Maybe we could add to spack core. We also use it when merging v5+v6

File: var/spack/repos/builtin/packages/neurodamus-neocortex/package.py (hunk 10 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-neocortex/package.py:25]: Oh, I thought we were building a single v5+v6 neurodamus, to behave like the previous neurodamus. But that's a idea I like, but then how so we setup PATH?

Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-neocortex/package.py:25]: I like the idea of a single module better. But if not possible, maybe?:

- neurodamus/neocortex-v5-0.1
- neurodamus/neocortex-v6-0.1

**Comment by ferdonline** [comment]
Ok, packages were updated to feature links and made the change here in the build.

**Review by pramodk — APPROVED** [review]
@ferdonline  : feel free to merge this now!

**Comment by ferdonline** [comment]
And we had LC_ALL biting again! :p 

All looking good now!!! 🎉
https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.SimulationStack/detail/hpc.SimulationStack/346/pipeline/

**Pull request merged**